Std/Data/String/Lemmas.lean

```diff
@@ -67,4 +72,9 @@ def utf8InductionOn {motive : List Char → Pos → Sort u}
       h ▸ eq c cs
     else ind c cs i h (utf8InductionOn cs (i + c) p nil eq ind)
 
+@[simp] theorem drop_empty {n : Nat} : "".drop n = "" := by
+  induction n with
+  | zero => rfl
+  | succ _ ih => exact ih
```

Comment:
Can this be proved by applying extensionality and then a list lemma? I'm sure there are many more like this.

Reply:
```lean
@[ext] theorem ext {s₁ s₂ : String} (h : s₁.data = s₂.data) : s₁ = s₂ :=
  show ⟨s₁.data⟩ = (⟨s₂.data⟩ : String) from h ▸ rfl

@[simp] theorem drop_empty {n : Nat} : "".drop n = "" := by
  apply ext
  show (drop "" n).data = []
  show [].drop n = ([] : List Char) -- Output: type mismatch
  exact List.drop_nil
```
@digama0 I don't think there's a way to do that.

Reply:
I don't think it's a `show`, that's a rewrite / `simp`

Reply:
```lean
ext] theorem ext {s₁ s₂ : String} (h : s₁.data = s₂.data) : s₁ = s₂ :=
  show ⟨s₁.data⟩ = (⟨s₂.data⟩ : String) from h ▸ rfl

theorem drop_eq : ∀ (s : String) (n : Nat), drop s n = ⟨s.toList.drop n⟩ := sorry

@[simp] theorem drop_empty {n : Nat} : "".drop n = "" := by
  apply ext
  rw [drop_eq]
  exact List.drop_nil
```
I used `String.drop_empty` to prove `String.drop_eq` in https://github.com/leanprover-community/mathlib4/pull/3789.

See https://github.com/leanprover-community/mathlib4/blob/5877ae2b8c35df2daf8ef993b937983edc1ea1e3/Mathlib/Data/String/Basic.lean#L172C1-L179.

Reply:
Can that theorem be here?

Reply:
> Can that theorem be here?

Is there a way to make the following proof work in Std? If so, I can move more lemmas from `Mathilb.Data.String.Basic` to here.
```lean
example (a b c : Nat) (h : a = b) : a + c = b + c := by
  simp
  exact h
```

Reply:
> (If it can't be used to prove `drop_empty` then so be it, but it should be part of the general simp strategy expressed by the supplied `@[simp]` lemmas.)
> 
> "I'll do it in the next PR" is also a valid response to this request.

It can't because its proof already uses `drop_empty`.  Do you want me to add the `[simp]` attribute to `drop_nil` and [`drop_cons_succ`](https://github.com/leanprover-community/mathlib4/blob/bd11b7ca87c813093af51effb060e41ef8305560/Mathlib/Data/String/Basic.lean#L165-L170)?

Reply:
> Is there a way to make the following proof work in Std?

I think it'd be hard to do so.

> `drop_eq`'s proof already uses `drop_empty`.

@digama0, will you merge this PR? I don't know any other way to prove `drop_empty` than using induction on the natural numbers.

Reply:
As I said, you don't need to change the proof of `drop_empty` but you should also put `drop_eq` (or rather, `toList_drop`) in this PR, unless there is a good reason not to, so that there is a confluent set of rewrites.

Reply:
@digama0 I moved lemmas including `drop_eq` from Mathlib to here.